api/alertmanager/api.go

```diff
@@ -0,0 +1,351 @@
+	}
+
+	return resp, body, err
+}
+
+// apiClient wraps a regular client and processes successful API responses.
+// Successful also includes responses that errored at the API level.
+type apiClient struct {
+	Client
+}
+
+type apiResponse struct {
+	Status    string          `json:"status"`
+	Data      json.RawMessage `json:"data"`
+	ErrorType ErrorType       `json:"errorType"`
+	Error     string          `json:"error"`
+}
+
+func (c apiClient) do(ctx context.Context, req *http.Request) (*http.Response, []byte, error) {
+	resp, body, err := c.Client.do(ctx, req)
+	if err != nil {
+		return resp, body, err
+	}
+
+	code := resp.StatusCode
+
+	if code/100 != 2 && code != statusAPIError {
+		return resp, body, &Error{
+			Type: ErrBadResponse,
+			Msg:  fmt.Sprintf("bad response code %d", resp.StatusCode),
+		}
+	}
+
+	var result apiResponse
+
+	if err = json.Unmarshal(body, &result); err != nil {
+		return resp, body, &Error{
+			Type: ErrBadResponse,
+			Msg:  err.Error(),
+		}
+	}
+
+	if (code == statusAPIError) != (result.Status == "error") {
+		err = &Error{
+			Type: ErrBadResponse,
+			Msg:  "inconsistent body for response code",
+		}
+	}
+
+	if code == statusAPIError && result.Status == "error" {
+		err = &Error{
+			Type: result.ErrorType,
+			Msg:  result.Error,
+		}
+	}
+
+	return resp, []byte(result.Data), err
+}
+
+type AlertAPI interface {
+	// Push a list of alerts into the Alertmanager.
+	Push(ctx context.Context, alerts ...*model.Alert) error
+}
+
+// NewAlertAPI returns a new AlertAPI for the client.
+func NewAlertAPI(c Client) AlertAPI {
+	return &httpAlertAPI{client: apiClient{c}}
+}
+
+type httpAlertAPI struct {
+	client Client
+}
+
+func (h *httpAlertAPI) Push(ctx context.Context, alerts ...*model.Alert) error {
+	u := h.client.url(epAlerts, nil)
+
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(alerts); err != nil {
+		return err
+	}
+
+	req, _ := http.NewRequest("POST", u.String(), &buf)
+
+	_, _, err := h.client.do(ctx, req)
+	return err
+}
+
+// SilenceAPI provides bindings the Alertmanager's silence API.
+type SilenceAPI interface {
+	// Get returns the silence associated with the given ID.
+	Get(ctx context.Context, id uint64) (*model.Silence, error)
+	// Set updates or creates the given silence and returns its ID.
+	Set(ctx context.Context, sil *model.Silence) (uint64, error)
+	// Del deletes the silence with the given ID.
+	Del(ctx context.Context, id uint64) error
+	// List all silences of the server.
+	List(ctx context.Context) ([]*model.Silence, error)
+}
+
+// NewSilenceAPI returns a new SilenceAPI for the client.
+func NewSilenceAPI(c Client) SilenceAPI {
+	return &httpSilenceAPI{client: apiClient{c}}
+}
+
+type httpSilenceAPI struct {
+	client Client
+}
+
+func (h *httpSilenceAPI) Get(ctx context.Context, id uint64) (*model.Silence, error) {
+	u := h.client.url(epSilence, map[string]string{
+		"id": strconv.FormatUint(id, 10),
+	})
+
+	req, _ := http.NewRequest("GET", u.String(), nil)
```

Comment:
Hmm, AFAICS it's even mandatory to act on the error in this situation because otherwise `req` will be `nil` and the subsequent `do` call will internally get `nil` pointer derefs if there's an error? Like in https://github.com/golang/net/blob/master/context/ctxhttp/cancelreq.go#L13